### ISSUE TITLE
FATAL EXCEPTION: buildingdata: building 2: warehouse: Unknown worker '[RANDOM CHARACTERS HERE]'

### ISSUE THREAD

**Issue by jasonkhanlar** [widelands/widelands#5501]
```
[00:00:13.406 real] ERROR: ##############################
[00:00:13.406 real] ERROR:   FATAL EXCEPTION: buildingdata: building 1: warehouse: Unknown worker 'x2'
[00:00:13.406 real] ERROR: ##############################
[00:00:13.406 real] ERROR:   Please report this problem to help us improve Widelands.
[00:00:13.406 real] ERROR:   You will find related messages in the standard output (stdout.txt on Windows).
[00:00:13.406 real] ERROR:   You are using build 1.0 (Release).
[00:00:13.406 real] ERROR:   Please add this information to your report.
[00:00:13.406 real] ERROR:   If desired, Widelands attempts to create an emergency savegame.
[00:00:13.406 real] ERROR:   It is often – though not always – possible to load it and continue playing.
[00:00:13.406 real] ERROR: ##############################

```

This error appears when trying to load a single player saved game that was originally a new random game.

Note: The line

```
 FATAL EXCEPTION: buildingdata: building 2: warehouse: Unknown worker '`@'
```

the characters in between the apostrophes are random/different for each attempt to reproduce this error (unique to each save file)

Steps to reproduce:

- Edit file /usr/share/widelands/tribes/initialization/barbarians/starting_conditions/headquarters.lua
  Change:
  ```
        soldiers = {
           [{0,0,0,0}] = 45,
        }
  ```
  to
  ```
        soldiers = {
           [{0,0,0,0}] = 65536,
        }
  ```
  Note: values 65535 and lower do not reproduce this error. I haven't tested for other workers and wares values, but they might be affected as well.
- Begin a new Single Player: New Random Game, make sure to select Barbarians as a tribe for at least 1 of the players, and start the game.
- Save the game.
- Exit the game.
- Load the saved game.
- You should see the error message above, both in console and dialog in game.

**Comment by jasonkhanlar** [comment]
I tested all the other values, and it appears that this error is only affecting the number of soldiers. I tried changing file:

/usr/share/widelands/tribes/initialization/barbarians/starting_conditions/headquarters.lua

to:

```
...
      wares = {
         ax = 65536,
         ax_battle = 65536,
         ax_broad = 65536,
         ax_bronze = 65536,
         ax_sharp = 65536,
         ax_warriors = 65536,
         barbarians_bread = 65536,
         beer = 65536,
         beer_strong = 65536,
         bread_paddle = 65536,
         blackwood = 65536,
         cloth = 65536,
         coal = 65536,
         felling_ax = 65536,
         fire_tongs = 65536,
         fish = 65536,
         fishing_rod = 65536,
         gold = 65536,
         gold_ore = 65536,
         granite = 65536,
         grout = 65536,
         hammer = 65536,
         helmet = 65536,
         helmet_mask = 65536,
         helmet_warhelm = 65536,
         hunting_spear = 65536,
         iron = 65536,
         iron_ore = 65536,
         kitchen_tools = 65536,
         log = 65536,
         meal = 65536,
         meat = 65536,
         pick = 65536,
         ration = 65536,
         reed = 65536,
         scythe = 65536,
         shovel = 65536,
         snack = 65536,
         water = 65536,
         wheat = 65536,
      },
      workers = {
         barbarians_baker = 65536,
         barbarians_blacksmith = 65536,
         barbarians_blacksmith_master = 65536,
         barbarians_brewer = 65536,
         barbarians_brewer_master = 65536,
         barbarians_builder = 65536,
         barbarians_carrier = 65536,
         barbarians_cattlebreeder = 65536,
         barbarians_charcoal_burner = 65536,
         barbarians_farmer = 65536,
         barbarians_ferry = 65536,
         barbarians_fisher = 65536,
         barbarians_gamekeeper = 65536,
         barbarians_gardener = 65536,
         barbarians_geologist = 65536,
         barbarians_helmsmith = 65536,
         barbarians_hunter = 65536,
         barbarians_innkeeper = 65536,
         barbarians_lime_burner = 65536,
         barbarians_lumberjack = 65536,
         barbarians_miner = 65536,
         barbarians_miner_chief = 65536,
         barbarians_miner_master = 65536,
         barbarians_ox = 65536,
         barbarians_ranger = 65536,
         barbarians_scout = 65536,
         barbarians_shipwright = 65536,
         barbarians_smelter = 65536,
         barbarians_stonemason = 65536,
         barbarians_trainer = 65536,
         barbarians_weaver = 65536,
      },
      soldiers = {
         [{0,0,0,0}] = 45,
      }
...
```

Note: I left soldiers data to default value, but tested all the others with 65536 (or higher) and going through the same steps described above, the error does not occur for any of the other units.

**Comment by Noordfrees** [comment]
When I try to add 70000 soldiers to my HQ via the debug console I get:
```
widelands: .../src/logic/player.cc:1428: void Widelands::Player::see_node(Widelands::MapIndex): Assertion `field.vision.seers() > 0' failed.
```

Such large numbers apparently cause overflows somewhere. Since I am very confident this can not happen in a regular game, best approach IMHO would be to check in Lua's `set_soldiers` that the number of soldiers won't exceed say 30000 (leaving a decent safety margin to `INT16_MAX`).

Thanks for finding this :)

**Comment by jmoerschbach** [comment]
this can be closed as via #5568 the limit was merged in set_soldiers I suppose @Noordfrees ?
